packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx

```diff
@@ -119,15 +119,15 @@ export class CustomHeadersSettings extends PureComponent<Props, State> {
       if (!header.configured) {
         secureJsonData[`httpHeaderValue${index + 1}`] = header.value;
       }
-      Object.keys(jsonData)
-        .filter(
-          (key) =>
-            key.startsWith('httpHeaderName') && parseInt(key.substring('httpHeaderName'.length), 10) > headers.length
-        )
-        .forEach((key) => {
-          delete jsonData[key];
-        });
     }
+    Object.keys(jsonData)
+      .filter(
+        (key) =>
+          key.startsWith('httpHeaderName') && parseInt(key.substring('httpHeaderName'.length), 10) > headers.length
+      )
+      .forEach((key) => {
+        delete jsonData[key];
+      });
```

Comment:
nit: I'm always a bit sceptical when I see deletes in a forEach, seems like we would mutate state. Maybe this is the same?
```suggestion
    const newJsonData: any = Object.keys(jsonData).reduce((acc, key) => {
      if (key.startsWith('httpHeaderName') && parseInt(key.substring('httpHeaderName'.length), 10) > headers.length) {
        return acc;
      }

      acc[key] = jsonData[key];
      return acc;
    }, {} as Record<string, any>);
```

Reply:
i'll be honest with you i was trying to do a minimal possible change here :-) you are right of course, i will adjust the code

Reply:
done (i chose a slightly different approach but should be ok)